In [4]:
import numpy as np

In [5]:
#创建一维数组，NumPy 中的一维数组没有明确的"行向量"和"列向量"之分
arr_1d = np.array([1, 2, 3])
print("一维数组:")
print(arr_1d)
print("shape:", arr_1d.shape)      # (3,) - 不是 (3,1) 也不是 (1,3)
print("ndim:", arr_1d.ndim)        # 1 - 一维
print("size:", arr_1d.size)        # 3

一维数组:
[1 2 3]
shape: (3,)
ndim: 1
size: 3


In [7]:
# 与二维数组对比
# 行向量 (1行, 3列)
row_vector = np.array([[1, 2, 3]])  # 注意双层括号
print("行向量 (1×3):")
print(row_vector)
print("shape:", row_vector.shape)  # (1, 3)
print("ndim:", row_vector.ndim)    # 2 - 二维！

# 列向量 (3行, 1列)
col_vector = np.array([[1], [2], [3]])
print("\n列向量 (3×1):")
print(col_vector)
print("shape:", col_vector.shape)  # (3, 1)
print("ndim:", col_vector.ndim)    # 2 - 二维！

行向量 (1×3):
[[1 2 3]]
shape: (1, 3)
ndim: 2

列向量 (3×1):
[[1]
 [2]
 [3]]
shape: (3, 1)
ndim: 2


In [9]:
#广播时的区别
# 创建矩阵
matrix = np.array([[1, 2, 3],
                   [4, 5, 6],
                   [7, 8, 9]])

print("矩阵 shape:", matrix.shape)  # (3, 3)

# 与一维数组相加
arr_1d = np.array([10, 20, 30])
result_1d = matrix + arr_1d
print("\n矩阵 + 一维数组:")
print(result_1d)
# [[11 22 33]
#  [14 25 36]
#  [17 28 39]]
# 广播规则：按行相加

# 与行向量相加
row_vec = np.array([[10, 20, 30]])  # shape: (1, 3)
result_row = matrix + row_vec
print("\n矩阵 + 行向量:")
print(result_row)
# [[11 22 33]
#  [14 25 36]
#  [17 28 39]]
# 与一维数组效果相同

# 与列向量相加
col_vec = np.array([[10], [20], [30]])  # shape: (3, 1)
result_col = matrix + col_vec
print("\n矩阵 + 列向量:")
print(result_col)
# [[11 12 13]
#  [24 25 26]
#  [37 38 39]]
# 按列相加，结果不同！

矩阵 shape: (3, 3)

矩阵 + 一维数组:
[[11 22 33]
 [14 25 36]
 [17 28 39]]

矩阵 + 行向量:
[[11 22 33]
 [14 25 36]
 [17 28 39]]

矩阵 + 列向量:
[[11 12 13]
 [24 25 26]
 [37 38 39]]


In [8]:
#矩阵乘法时的区别
# 矩阵乘法示例
# @ 操作符等价于 np.dot也等价于np.matmul
A = np.array([[1, 2, 3],
              [4, 5, 6]])  # 2×3

# 一维数组
v_1d = np.array([1, 2, 3])  # shape: (3,)
try:
    result = A @ v_1d  # 矩阵乘法
    print("A @ v_1d:", result)  # [14 32]
except Exception as e:
    print(f"错误: {e}")

# 列向量（二维）
v_col = np.array([[1], [2], [3]])  # shape: (3, 1)
result_col = A @ v_col
print("A @ v_col:")
print(result_col)
# [[14]
#  [32]]

# 行向量（二维）
v_row = np.array([[1, 2, 3]])  # shape: (1, 3)
try:
    result = A @ v_row.T  # 需要转置
    print("A @ v_row.T:")
    print(result)
except Exception as e:
    print(f"错误: {e}")

A @ v_1d: [14 32]
A @ v_col:
[[14]
 [32]]
A @ v_row.T:
[[14]
 [32]]


In [10]:
# 当 @ 左边是矩阵，右边是一维数组时：
# 一维数组被当作列向量
# 结果会被压缩（去掉大小为1的维度）
M = np.array([[1, 2], [3, 4]])  # 2×2
v = np.array([1, 2])            # (2,)

result = M @ v  # v被当作 (2, 1)
print("M @ v:", result)  # [ 5 11]
# 计算：1×1 + 2×2 = 5, 3×1 + 4×2 = 11

M @ v: [ 5 11]


In [11]:
# 当 @ 左边是一维数组，右边是矩阵时：
# 一维数组被当作行向量
# 结果会被压缩（去掉大小为1的维度）
v = np.array([1, 2])            # (2,)
M = np.array([[1, 2], [3, 4]])  # 2×2

result = v @ M  # v被当作 (1, 2)
print("v @ M:", result)  # [ 7 10]
# 计算：1×1 + 2×3 = 7, 1×2 + 2×4 = 10

v @ M: [ 7 10]


In [12]:
# 一维数组 ↔ 二维数组
# 创建一维数组
arr_1d = np.array([1, 2, 3])

# 转换为行向量 (1×3)
row_vector = arr_1d[np.newaxis, :]  # 或 arr_1d[None, :]
print("转换为行向量:")
print(row_vector)
print("shape:", row_vector.shape)  # (1, 3)

# 转换为列向量 (3×1)
col_vector = arr_1d[:, np.newaxis]  # 或 arr_1d[:, None]
print("\n转换为列向量:")
print(col_vector)
print("shape:", col_vector.shape)  # (3, 1)

# reshape方法
row_vec2 = arr_1d.reshape(1, -1)  # -1表示自动计算
col_vec2 = arr_1d.reshape(-1, 1)
print("\nreshape方法:")
print("行向量:", row_vec2.shape)
print("列向量:", col_vec2.shape)

# expand_dims函数
row_vec3 = np.expand_dims(arr_1d, axis=0)  # 在第0维增加维度
col_vec3 = np.expand_dims(arr_1d, axis=1)  # 在第1维增加维度
print("\nexpand_dims函数:")
print("行向量:", row_vec3.shape)
print("列向量:", col_vec3.shape)

转换为行向量:
[[1 2 3]]
shape: (1, 3)

转换为列向量:
[[1]
 [2]
 [3]]
shape: (3, 1)

reshape方法:
行向量: (1, 3)
列向量: (3, 1)

expand_dims函数:
行向量: (1, 3)
列向量: (3, 1)


In [13]:
#线性代数运算

# 点积（内积）
a = np.array([1, 2, 3])
b = np.array([4, 5, 6])

# 一维数组的点积
dot_1d = np.dot(a, b)
print("一维数组点积:", dot_1d)  # 32 (1*4 + 2*5 + 3*6)

# 二维数组的矩阵乘法
a_row = a.reshape(1, -1)  # 行向量
b_col = b.reshape(-1, 1)  # 列向量
dot_2d = a_row @ b_col    # 矩阵乘法
print("二维数组矩阵乘法:")
print(dot_2d)  # [[32]]

一维数组点积: 32
二维数组矩阵乘法:
[[32]]


In [14]:
#统计计算
data = np.array([[1, 2, 3],
                 [4, 5, 6],
                 [7, 8, 9]])

# 一维数组的mean
arr_1d = np.array([1, 2, 3, 4, 5])
print("一维数组平均值:", arr_1d.mean())  # 标量

# 二维数组按轴求平均
print("\n二维数组平均值:")
print("所有元素平均:", data.mean())          # 标量
print("按列平均 (axis=0):", data.mean(axis=0))  # 行向量？实际是一维数组
print("按行平均 (axis=1):", data.mean(axis=1))  # 一维数组

# 注意：axis运算默认返回一维数组
col_means = data.mean(axis=0)
print("col_means shape:", col_means.shape)  # (3,)，不是 (1,3)

一维数组平均值: 3.0

二维数组平均值:
所有元素平均: 5.0
按列平均 (axis=0): [4. 5. 6.]
按行平均 (axis=1): [2. 5. 8.]
col_means shape: (3,)


In [16]:
matrix = np.array([[1,2],
                   [1,2],
                   [1,2]])
print(matrix)

[[1 2]
 [1 2]
 [1 2]]


In [17]:
matrix_object = np.mat([[1,2],
                         [1,2],
                         [1,2]])
print(matrix_object)

[[1 2]
 [1 2]
 [1 2]]


In [2]:
from scipy import sparse

In [20]:
matrix = np.array([[0,0],
                  [0,1],
                  [3,0]]) 
matrix_sparse = sparse.csr_matrix(matrix)  
print(matrix_sparse)                    

<Compressed Sparse Row sparse matrix of dtype 'int32'
	with 2 stored elements and shape (3, 2)>
  Coords	Values
  (1, 1)	1
  (2, 0)	3


In [21]:
matrix_large = np.array([[0,0,0,0,0,0,0,0,0,0],
                         [0,1,0,0,0,0,0,0,0,0],
                         [3,0,0,0,0,0,0,0,0,0],])
matrix_large_sparse = sparse.csr_matrix(matrix_large)
print(matrix_large_sparse)

<Compressed Sparse Row sparse matrix of dtype 'int32'
	with 2 stored elements and shape (3, 10)>
  Coords	Values
  (1, 1)	1
  (2, 0)	3


In [22]:
# 创建稠密矩阵
matrix_large = np.array([
    [0,0,0,0,0,0,0,0,0,0],
    [0,1,0,0,0,0,0,0,0,0],
    [3,0,0,0,0,0,0,0,0,0],
])

# 转换为稀疏矩阵
matrix_large_sparse = sparse.csr_matrix(matrix_large)

print("稠密矩阵内存:", matrix_large.nbytes, "bytes")
print("稀疏矩阵内存:", matrix_large_sparse.data.nbytes + 
                     matrix_large_sparse.indices.nbytes + 
                     matrix_large_sparse.indptr.nbytes, "bytes")

# 更大规模的对比
large_dense = np.zeros((1000, 1000))
large_dense[10, 20] = 1
large_dense[50, 100] = 2
large_dense[200, 300] = 3

large_sparse = sparse.csr_matrix(large_dense)

print(f"\n1000×1000矩阵（只有3个非零元素）:")
print(f"稠密矩阵: {large_dense.nbytes:,} bytes")      # 8,000,000 bytes
print(f"稀疏矩阵: {(large_sparse.data.nbytes + large_sparse.indices.nbytes + large_sparse.indptr.nbytes):,} bytes")  # 约 100 bytes
print(f"节省了 {large_dense.nbytes/(large_sparse.data.nbytes + large_sparse.indices.nbytes + large_sparse.indptr.nbytes):.0f}倍内存！")

稠密矩阵内存: 120 bytes
稀疏矩阵内存: 32 bytes

1000×1000矩阵（只有3个非零元素）:
稠密矩阵: 8,000,000 bytes
稀疏矩阵: 4,040 bytes
节省了 1980倍内存！


In [3]:
import numpy as np
from scipy import sparse
import time

# 创建大型稀疏矩阵
size = 20000
density = 0.001  # 0.1% 非零元素

# 随机稀疏矩阵
sparse_mat = sparse.random(size, size, density=density, format='csr')
dense_mat = sparse_mat.toarray()

# 向量乘法性能对比
vector = np.random.randn(size)

# 稀疏矩阵乘法
start = time.time()
result_sparse = sparse_mat @ vector
time_sparse = time.time() - start

# 稠密矩阵乘法  
start = time.time()
result_dense = dense_mat @ vector
time_dense = time.time() - start

print(f"矩阵大小: {size}×{size}")
print(f"非零元素比例: {density*100:.1f}%")
print(f"稀疏矩阵乘法时间: {time_sparse:.4f}s")
print(f"稠密矩阵乘法时间: {time_dense:.4f}s")
print(f"稀疏矩阵快 {time_dense/time_sparse:.1f}倍")

矩阵大小: 20000×20000
非零元素比例: 0.1%
稀疏矩阵乘法时间: 0.0010s
稠密矩阵乘法时间: 3.4056s
稀疏矩阵快 3397.7倍


In [6]:
import numpy as np

vector = np.array([1, 2, 3, 4, 5])
matrix = np.array([[1,2,3],
                   [4,5,6],
                   [7,8,9]])
print(vector[2])

print(matrix[1, 1])

3
5


In [8]:
print(vector[:])

[1 2 3 4 5]


In [9]:
print(vector[:3])

[1 2 3]


In [10]:
vector[3:]

array([4, 5])

In [11]:
vector[-1]

5

In [12]:
matrix[:2,:]

array([[1, 2, 3],
       [4, 5, 6]])

In [14]:
print(matrix[:,1:2])

[[2]
 [5]
 [8]]


In [15]:
matrix = np.array([[1,2,3,4],
                     [5,6,7,8],
                     [9,10,11,12]])
matrix.shape

(3, 4)

In [16]:
matrix.size

12

In [17]:
matrix.ndim

2

In [23]:
matrix = np.array([[1,2,3],
                     [4,5,6],
                     [7,8,9]])

add_100 = lambda x: x + 100
#np.vectorize（伪向量化）
vector_add_100 = np.vectorize(add_100)
result = vector_add_100(matrix)
result

array([[101, 102, 103],
       [104, 105, 106],
       [107, 108, 109]])

In [24]:
matrix + 100

array([[101, 102, 103],
       [104, 105, 106],
       [107, 108, 109]])

In [20]:
import numpy as np
arr = np.array([1.0, -2.0, 3.0, 0.0, -1.0, 2.0])

result = np.zeros_like(arr)
mask = arr > 0
#当使用 arr[mask] 从数组中读取数据时，会创建一个新的数组，包含所有满足条件的元素。
#当使用 result[mask] = values 向数组写入数据时，会直接修改原数组的指定位置。
result[mask] = np.sin(arr[mask])    # 向量化
result[~mask] = np.cos(arr[~mask])  # 向量化
result

array([ 0.84147098, -0.41614684,  0.14112001,  1.        ,  0.54030231,
        0.90929743])

In [21]:
# 当使用切片时，情况稍有不同
arr = np.array([1, 2, 3, 4, 5])
view = arr[:3]        # 创建视图（view）
view[0] = 100         # 会修改原数组！
print(arr)            # [100, 2, 3, 4, 5]

# 但布尔索引通常返回副本
arr = np.array([1, 2, 3, 4, 5])
mask = [True, True, True, False, False]
selected = arr[mask]  # 这是副本
selected[0] = 200     # 不会影响原数组
print(arr)            # [1, 2, 3, 4, 5]

[100   2   3   4   5]
[1 2 3 4 5]


In [22]:
import numpy as np

# 创建原数组
arr = np.array([1, 2, 3, 4, 5])
print("原数组:", arr)           # [1 2 3 4 5]
print("原数组id:", id(arr))

# 创建视图（通过切片）
view = arr[:3]                  # 前3个元素
print("\n视图:", view)           # [1 2 3]
print("视图id:", id(view))
print("view.base is arr?", view.base is arr)  # True

# 修改视图
view[0] = 100
print("\n修改视图后:")
print("视图:", view)             # [100 2 3]
print("原数组:", arr)            # [100 2 3 4 5] 也被修改了！

原数组: [1 2 3 4 5]
原数组id: 1873101717552

视图: [1 2 3]
视图id: 1873081252432
view.base is arr? True

修改视图后:
视图: [100   2   3]
原数组: [100   2   3   4   5]


In [26]:
matrix = np.array([[1, 2, 3],
                     [4, 5, 6],
                     [7, 8, 9]])
print(np.max(matrix, axis=0))  # 按列取最大值
print(np.max(matrix, axis=1))  # 按行取最大值
print(np.max(matrix))         # 全局最大值

[7 8 9]
[3 6 9]
9


In [28]:
np.mean(matrix, axis=0)  # 按列求平均值


array([4., 5., 6.])

In [29]:
np.mean(matrix, axis=1)  # 按行求平均值


array([2., 5., 8.])

In [30]:
np.mean(matrix)         # 全局平均值

5.0

In [31]:
np.var(matrix)

6.666666666666667

In [32]:
np.std(matrix)

2.581988897471611

In [33]:
matrix = np.array([[1, 2, 3],
                    [4, 5, 6],
                    [7, 8, 9],
                    [10, 11, 12]])
matrix.reshape(2,6)

array([[ 1,  2,  3,  4,  5,  6],
       [ 7,  8,  9, 10, 11, 12]])

In [34]:
matrix.reshape(1,-1)

array([[ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12]])

In [35]:
matrix.reshape(12)

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12])

In [36]:
matrix.T

array([[ 1,  4,  7, 10],
       [ 2,  5,  8, 11],
       [ 3,  6,  9, 12]])

In [37]:
np.array([1, 2, 3, 4, 5, 6]).T

array([1, 2, 3, 4, 5, 6])

In [38]:
np.array([[1, 2, 3, 4, 5 ,6]]).T

array([[1],
       [2],
       [3],
       [4],
       [5],
       [6]])

In [39]:
matrix.flatten()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12])

In [40]:
matrix

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [10, 11, 12]])

In [41]:
matrix = np.array([[1, 1, 1],
                   [1, 1, 10],
                   [1, 1, 15]])

np.linalg.matrix_rank(matrix)

2

In [43]:
matrix = np.array([[1,2,3],
                   [2,4,6],
                   [3,8,9]])
np.linalg.det(matrix)

0.0

In [44]:
matrix.diagonal()

array([1, 4, 9])

In [45]:
matrix.diagonal(offset=1)

array([2, 6])

In [46]:
matrix.diagonal(offset=-1)

array([2, 8])

In [47]:
matrix.trace()

14

In [48]:
sum(matrix.diagonal())

14

In [49]:
matrix = np.array([[1, -1, 3],
                    [1, 1, 6],
                    [3, 8, 9]])
eigenvalues, eigenvectors = np.linalg.eig(matrix)

In [50]:
eigenvalues

array([13.55075847,  0.74003145, -3.29078992])

In [51]:
eigenvectors

array([[-0.17622017, -0.96677403, -0.53373322],
       [-0.435951  ,  0.2053623 , -0.64324848],
       [-0.88254925,  0.15223105,  0.54896288]])

In [52]:
vector_a = np.array([1, 2, 3])
vector_b = np.array([4, 5, 6])
np.dot(vector_a, vector_b)

32

In [53]:
vector_a @ vector_b

32

In [55]:
matrix_a = np.array([[1, 1, 1],
 [1, 1, 1],
 [1, 1, 2]])
matrix_b = np.array([[1, 3, 1],
 [1, 3, 1],
 [1, 3, 8]])

In [56]:
np.add(matrix_a, matrix_b)

array([[ 2,  4,  2],
       [ 2,  4,  2],
       [ 2,  4, 10]])

In [57]:
np.subtract(matrix_a, matrix_b)

array([[ 0, -2,  0],
       [ 0, -2,  0],
       [ 0, -2, -6]])

In [58]:
matrix_a + matrix_b

array([[ 2,  4,  2],
       [ 2,  4,  2],
       [ 2,  4, 10]])

In [59]:
matrix_a = np.array([[1, 1],
 [1, 2]])
matrix_b = np.array([[1, 3],
 [1, 2]])

In [60]:
np.dot(matrix_a, matrix_b)

array([[2, 5],
       [3, 7]])

In [61]:
matrix_a @ matrix_b

array([[2, 5],
       [3, 7]])

In [63]:
#对应元素相乘（Hadamard积）
matrix_a * matrix_b

array([[1, 3],
       [1, 4]])

In [66]:
matrix = np.array([[1, 4],
                    [2, 5],
                    ]
)
np.linalg.inv(matrix)


array([[-1.66666667,  1.33333333],
       [ 0.66666667, -0.33333333]])

In [67]:
matrix @ np.linalg.inv(matrix)

array([[1., 0.],
       [0., 1.]])

In [78]:
np.random.seed(0)
np.random.random(3)

array([0.5488135 , 0.71518937, 0.60276338])

In [76]:
np.random.random(3)

array([0.54488318, 0.4236548 , 0.64589411])

In [79]:
np.random.randint(0,11,3)

array([3, 7, 9])

In [80]:
np.random.normal(0, 1, 3 )

array([-1.42232584,  1.52006949, -0.29139398])

In [81]:
np.random.logistic(0, 1, 3)

array([-0.98118713, -0.08939902,  1.46416405])

In [82]:
np.random.uniform(1,2,3)

array([1.47997717, 1.3927848 , 1.83607876])

In [83]:
from sklearn import datasets

In [84]:
digits = datasets.load_digits()

In [95]:
features = digits.data
features.shape

(1797, 64)

In [97]:
target = digits.target
target[0]

0

In [96]:
features[0]

array([ 0.,  0.,  5., 13.,  9.,  1.,  0.,  0.,  0.,  0., 13., 15., 10.,
       15.,  5.,  0.,  0.,  3., 15.,  2.,  0., 11.,  8.,  0.,  0.,  4.,
       12.,  0.,  0.,  8.,  8.,  0.,  0.,  5.,  8.,  0.,  0.,  9.,  8.,
        0.,  0.,  4., 11.,  0.,  1., 12.,  7.,  0.,  0.,  2., 14.,  5.,
       10., 12.,  0.,  0.,  0.,  0.,  6., 13., 10.,  0.,  0.,  0.])

digits 数据集是什么？
包含 1797 张 8×8 像素的手写数字图像

每个像素的灰度值（0-16）就是一个特征(包括0，不包括16)

总共 8×8 = 64 个特征